### PR TITLE
[Replicate] Simplify code a bit

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98890
* __->__ #98889

Simplifies the code, such as making self.modules not a list and only a
single module.

Differential Revision: [D44899281](https://our.internmc.facebook.com/intern/diff/D44899281/)

### CODE REVIEW

File: torch/distributed/_composable/replicate.py [hunk 1]
```diff
@@ -34,23 +34,21 @@ def _can_compose(module: nn.Module) -> bool:
 
 class _ReplicateState:
     def __init__(self) -> None:
-        self.modules: List[nn.Module] = []
+        self.module = None
```
Comment (mrshenli) [inline torch/distributed/_composable/replicate.py:37]: nit: keep `nn.Module` typing?

**Review by mrshenli — APPROVED** [review]
LGTM. Please fix lint before landing :)

**Review by yhcharles — APPROVED** [review]
Thanks, LGTM with a minor comment

**Pull request closed without merging**